### ISSUE TITLE
Refactor Request: PregnancyProgression.as

### ISSUE THREAD

**Issue by misplaced-radish** [Kitteh6660/Corruption-of-Champions-Mod#322]
I don't have any experience with actionscript and I'm leery of doing any major changes for obvious reasons, but if anyone has time, would they mind fixing up classes/classes/Scenes/PregnancyProgression.as to use functions and be more readable?  On a glance I see the lactation increases ought to be straightforward.

**Comment by anonymph** [comment]
Fucking hell, that is bad...

I'll take a shot at it. Probably over a few rounds...

Thinking:
1. Segment the current function properly, and replace dumb `if` statements with full `if...else if...else`, where relevant.
2. Move related sections into either own (`private` scoped) functions.
3. Do eventual additional restructuring.

**Comment by misplaced-radish** [comment]
That all sounds good!  Currently the pregnancies moved off into character files are the ones with their own lengthy scenes, while the brief updates are in PregnancyProgression, which makes sense to me; they just really ought to all be functions and call the lactation stuff.  The lactation stuff looks like it all works about the same, but they have different ideals for how big the breasts ought to get.

Yes, you can be pregnant through both holes.  Note also that bunny eggs can be through either; they're currently copy-pasted between the vaginal pregnancy if statements and the butt pregnancy ones, nowhere near each other.

**Comment by Stadler76** [comment]
mmh
```as3
			if (player.pregnancyIncubation > 0 && player.pregnancyIncubation < 2) player.knockUpForce(player.pregnancyType, 1);
```
... ouch!

**Comment by Stadler76** [comment]
As a first step (one or multiple commits) I'd suggest to move all those checks for
```as3
if (player.pregnancyIncubation == 1 && player.pregnancyType == PregnancyStore.PREGNANCY_FAERIE) { /* ... */ }
// ...
if (player.pregnancyType == PregnancyStore.PREGNANCY_URTA && player.pregnancyIncubation == 1)  { /* ... */ }
```
and so on into
`if (player.pregnancyIncubation == 1) {`
and do the same for 
`if (player.buttPregnancyIncubation == 1) {`

Makes collapsing code easier, when you can simply collapse these values.
IIRC `player.pregnancyIncubation` and `player.buttPregnancyIncubation` stores the remaining hours, till you give birth/lay eggs.

In the **next seperate** step (one or two commits, I suppose) I'd suggest to change this to use a switch-case. IIRC all of them look up the `player.(buttP|p)regnancyType` value.
Note, that you can start with a few `case` es and leave the rest of the if's in the `default:` section of the switch-case

**Comment by anonymph** [comment]
Already pretty far ahead. Currently grouping them in functions per `pregnancyType` as that allows one to read through an entire pregnancy in one go. Eventual util functions are being implemented as I get annoying fixing the same lines of code.

One thing I've noticed is that some of the "migrated" pregnancies contains sniplets of code that I'm moving into util functions, and I'm not sure if I should pull the migrated content back into `PregnancyProgression.as`, or make the util functions `public`, or not bother.

Personally I hate C-style switch statements, due to the `break` keyword. I'll probably implement selection of the relevant function by using an array with references to the functions, now that I'm already dividing them in that way.

**Comment by Stadler76** [comment]
Yeah, but be aware, that there may be differences in texts depending on buttPregnancyIncubation and/or pregnancyIncubation for a reason. So you shouldn't blindly merge those texts. 

> Personally I hate C-style switch statements, due to the break keyword.

In other words: Your personal preference is more important, than creating human readable code?
I know, that forgetting a break; can cause unwanted behaviour, but be aware, that if you break things, by not `break;`ing things, I'll be there to fix it by `break;`ing it correctly. ;-)

**Comment by anonymph** [comment]
I've separated butt and vagina based pregnancies into different function (except for Satyrs where they were exactly the same), don't worry.

Well, my personal preference is that it's more readable `;)`. C-style switch statements really hurt readability by requiring `break`, but I get your point. I'll attempt to make it readable to others.

**Comment by Stadler76** [comment]
Try this (took me some time and lots of try&error to get this working):
```diff
diff --git a/classes/classes/Items/MutationsHelper.as b/classes/classes/Items/MutationsHelper.as
index c2cda5b..824501c 100644
--- a/classes/classes/Items/MutationsHelper.as
+++ b/classes/classes/Items/MutationsHelper.as
@@ -1,6 +1,8 @@
 package classes.Items 
 {
 	import classes.*;
+	import mx.utils.SHA256;
+	import flash.utils.ByteArray;
 	
 	/**
 	 * Helper class to get rid of the copy&paste-mess in classes.Items.Mutations
@@ -220,6 +222,11 @@ package classes.Items
 		public function updateOvipositionPerk(tfSource:String):int
 
 		{
+			var tfSourceBA:ByteArray = new ByteArray;
+			var tfSourceHash:String;
+			tfSourceBA.writeUTFBytes(tfSource);
+			tfSourceHash = SHA256.computeDigest(tfSourceBA);
+			trace('called updateOvipositionPerk("' + tfSourceHash + '")');
 			trace('called updateOvipositionPerk("' + tfSource + '")');
 			var tsParts:Array = tfSource.split("-");
 			if (tsParts.length > 1 && ["goldenSeed", "catTransformation"].indexOf(tsParts[0]) == -1)

```

I'll try to add sha256(str:string) to Utils.as in a whiff.

**Comment by Stadler76** [comment]
Here we go: Tested and results in the same as PHP-code:
```php
echo hash('SHA256', $string);
```

The diff for CoC Revamp:
```diff
diff --git a/classes/classes/Items/MutationsHelper.as b/classes/classes/Items/MutationsHelper.as
index c2cda5b..e299106 100644
--- a/classes/classes/Items/MutationsHelper.as
+++ b/classes/classes/Items/MutationsHelper.as
@@ -1,6 +1,7 @@
 package classes.Items 
 {
 	import classes.*;
+	import classes.internals.Utils;
 	
 	/**
 	 * Helper class to get rid of the copy&paste-mess in classes.Items.Mutations
@@ -218,8 +219,8 @@ package classes.Items
 		 * @author	Stadler76
 		 */
 		public function updateOvipositionPerk(tfSource:String):int
-
 		{
+			trace('called updateOvipositionPerk("' + Utils.sha256(tfSource) + '")');
 			trace('called updateOvipositionPerk("' + tfSource + '")');
 			var tsParts:Array = tfSource.split("-");
 			if (tsParts.length > 1 && ["goldenSeed", "catTransformation"].indexOf(tsParts[0]) == -1)
diff --git a/classes/classes/internals/Utils.as b/classes/classes/internals/Utils.as
index 90a205d..b6f2012 100644
--- a/classes/classes/internals/Utils.as
+++ b/classes/classes/internals/Utils.as
@@ -4,6 +4,9 @@
 package classes.internals
 {
 	import classes.*;
+	import mx.utils.SHA256;
+	import flash.utils.ByteArray;
+
 	public class Utils extends Object
 	{
 		private static const NUMBER_WORDS_NORMAL:Array		= ["zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine", "ten"];
@@ -13,7 +16,14 @@ package classes.internals
 		public function Utils()
 		{
 		}
-		
+
+		public static function sha256(str:String):String
+		{
+			var strBytes:ByteArray = new ByteArray;
+			strBytes.writeUTFBytes(str);
+			return SHA256.computeDigest(strBytes);
+		}
+
 		// curryFunction(f,args1)(args2)=f(args1.concat(args2))
 		// e.g. curryFunction(f,x,y)(z,w) = f(x,y,z,w)
 		public static function curry(func:Function,...args):Function
```

**Comment by Stadler76** [comment]
Bump. Anyone brave enough to look into this soonish? The whole file is a mess.

**Comment by misplaced-radish** [comment]
I'm still too terrified to try fixing it myself, but I'm glad it hasn't been totally forgotten!

I got a question: would making the bunny egg preg a status effect instead of a pregnancy that works entirely different from every other pregnancy in there be a good idea?  I'm pretty sure that's the only one that does something different than "check if it's time to post a message and if the message has been posted, post the message, and adjust your breast size + lactation."

**Comment by brrritssocold** [comment]
I'm willing to take a stab at this, looked over the code and did some scratch refactoring to get a sense of how the class works.

Yeah, cleaning this up will take considerable effort, but should be worth it in the long run.
Will make a series of PRs to chip away at the class, bit by bit.
